### ISSUE TITLE
msiSetDefaultResc not being honoured on destination Federated Zones?

### ISSUE THREAD

**Issue by kript** [irods/irods#3224]
Might not be just msiSetDefaultResc, but using as a demo;

precis;
1. rules on federated zone set one compound resource tree on one collection, and another on all other collections, via msiSetDefaultResc (see rules below). 
2. Federated user uploads a file via iput, connecting to federation master, file ends up in demoResc on iCAT of destination zone.
3. local user uploads file to same collection, rules are honored.

from client as Federated user
```
time iput -K vmstat.gluplot /seq-dev/jc18_testcol/vmstat.gluplot.7

real	0m2.219s
user	0m0.020s
sys	0m0.020s

 ils -L /seq-dev/jc18_testcol/vmstat.gluplot.7
  jc18              0 demoResc         1045 2016-07-12.16:40 & vmstat.gluplot.7
    558ef7303de65254106e6a65bf8ed02d    generic    /usr/local/iRODS/Vault/jc18_testcol/vmstat.gluplot.7

 ienv | grep resource
NOTICE: irods_default_resource is not defined
```

```
iuserinfo jc18
...
name: jc18
id: 12085
type: rodsuser
zone: Sanger1-dev
info: 
comment: 
create time: 01458052529: 2016-03-15.14:35:29
modify time: 01458052529: 2016-03-15.14:35:29
member of group: jc18
member of group: public
```

server_config.json;
```
    "re_rulebase_set": [
        {
            "filename": "acAclPolicy"
        },
        {
            "filename": "seq-dev"
        },
        {
            "filename": "core"
        }
        
    ], 
```

seq-dev.re file
```
acSetRescSchemeForCreate {
        if ($objPath like regex '/seq-dev/illumina/library_merge/.*\.(bam|cram)$') {
                msiSetDefaultResc("noReplRoot","null");
		writeLine("serverLog", "writing to noReplRoot");
        }
        else {
                msiSetDefaultResc("root","null");
		writeLine("serverLog", "writing to root");
        }
}
acPreprocForDataObjOpen{ msiSetDataObjPreferredResc("root"); }
  
acPostProcForPut {msiSysChksumDataObj; }
```

Federation master logs;
```
Jul 12 16:40:08 pid:32207 NOTICE: Agent process 16975 started for puser=jc18 and cuser=jc18 from 172.17.122.10
Jul 12 16:40:08 pid:16975 NOTICE: rsAuthResponse set proxy authFlag to 3, client authFlag to 3, user:jc18#Sanger1-dev proxy:jc18 client:jc18
Jul 12 16:40:09 pid:32207 NOTICE: Agent process 16976 started for puser=irods-g1-dev and cuser=jc18 from 172.XXX.XXX.XXX
Jul 12 16:40:09 pid:16976 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:09 pid:16976 NOTICE: Agent exiting with status = 0
Jul 12 16:40:09 pid:32207 NOTICE: Agent process 16976 exited with status 0
Jul 12 16:40:09 pid:32207 NOTICE: Agent process 16977 started for puser=irods-g1-dev and cuser=jc18 from 172.XXX.XXX.XXX
Jul 12 16:40:09 pid:16977 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:09 pid:16977 NOTICE: Agent exiting with status = 0
Jul 12 16:40:09 pid:32207 NOTICE: Agent process 16977 exited with status 0
Jul 12 16:40:09 pid:32207 NOTICE: Agent process 16978 started for puser=irods-g1-dev and cuser=jc18 from 172.XXX.XXX.XXX
Jul 12 16:40:09 pid:16978 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:09 pid:16978 NOTICE: Agent exiting with status = 0
Jul 12 16:40:10 pid:32207 NOTICE: Agent process 16978 exited with status 0
Jul 12 16:40:10 pid:32207 NOTICE: Agent process 16987 started for puser=irods-g1-dev and cuser=jc18 from 172.XXX.XXX.XXX
Jul 12 16:40:10 pid:16987 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:10 pid:16987 NOTICE: Agent exiting with status = 0
Jul 12 16:40:10 pid:16975 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:11 pid:16975 NOTICE: Agent exiting with status = 0
Jul 12 16:40:11 pid:32207 NOTICE: Agent process 16975 exited with status 0
Jul 12 16:40:11 pid:32207 NOTICE: Agent process 16987 exited with status 0
Jul 12 16:40:18 pid:32207 NOTICE: Agent process 16988 started for puser=jc18 and cuser=jc18 from 172.XXX.XXX.XXX
Jul 12 16:40:18 pid:16988 NOTICE: rsAuthResponse set proxy authFlag to 3, client authFlag to 3, user:jc18#Sanger1-dev proxy:jc18 client:jc18
Jul 12 16:40:19 pid:32207 NOTICE: Agent process 16989 started for puser=irods-g1-dev and cuser=jc18 from 172.XXX.XXX.XXX
Jul 12 16:40:19 pid:16989 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:19 pid:16989 NOTICE: Agent exiting with status = 0
Jul 12 16:40:19 pid:32207 NOTICE: Agent process 16989 exited with status 0
Jul 12 16:40:19 pid:32207 NOTICE: Agent process 16990 started for puser=irods-g1-dev and cuser=jc18 from 172.XXX.XXX.XXX
Jul 12 16:40:19 pid:16990 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:19 pid:16990 NOTICE: Agent exiting with status = 0
Jul 12 16:40:19 pid:32207 NOTICE: Agent process 16990 exited with status 0
Jul 12 16:40:19 pid:32207 NOTICE: Agent process 16991 started for puser=irods-g1-dev and cuser=jc18 from 172.XXX.XXX.XXX
Jul 12 16:40:19 pid:16991 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:19 pid:16991 NOTICE: Agent exiting with status = 0
Jul 12 16:40:19 pid:16988 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:19 pid:16988 NOTICE: Agent exiting with status = 0
Jul 12 16:40:19 pid:32207 NOTICE: Agent process 16988 exited with status 0
Jul 12 16:40:19 pid:32207 NOTICE: Agent process 16991 exited with status 0
```

log of iCAT on federated Zone;
```
Jul 12 16:40:08 pid:9183 NOTICE: Agent process 10241 started for puser=irods-dev and cuser=jc18 from 172.XXX.XXX.XXX
Jul 12 16:40:09 pid:10241 NOTICE: writeLine: inString = writing to root
Jul 12 16:40:09 pid:9183 NOTICE: Agent process 10242 started for puser=irods-g1-dev and cuser=jc18 from 172.XXX.XXX.XXX
Jul 12 16:40:10 pid:10241 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:10 pid:10241 NOTICE: Agent exiting with status = 0
Jul 12 16:40:10 pid:10242 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:10 pid:10242 NOTICE: Agent exiting with status = 0
Jul 12 16:40:11 pid:9183 NOTICE: Agent process 10241 exited with status 0
Jul 12 16:40:11 pid:9183 NOTICE: Agent process 10242 exited with status 0
Jul 12 16:40:18 pid:9183 NOTICE: Agent process 10243 started for puser=irods-dev and cuser=jc18 from 172.XXX.XXX.XXX
Jul 12 16:40:19 pid:10243 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 12 16:40:19 pid:10243 NOTICE: Agent exiting with status = 0
Jul 12 16:40:19 pid:9183 NOTICE: Agent process 10243 exited with status 0
```

this log was generated on one of the iRES;

```
Jul 13 11:27:39 pid:15399 NOTICE: Agent process 24030 started for puser=irods-dev and cuser=jc18 from 172.XXX.XXX.XXX
Jul 13 11:27:39 pid:24030 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 13 11:27:39 pid:24030 NOTICE: Agent exiting with status = 0
Jul 13 11:27:40 pid:15399 NOTICE: Agent process 24030 exited with status 0
```

but not on the other...

Note that demoResc is the default resource for the environment for the user running iRODS on the iCAT of the Federated Zone;

```
grep resource /usr/local/iRODS/.irods/irods_environment.json
    "irods_default_resource": "demoResc",
```

If you use a local to the zone user, the rules are honored;

```
$ iput -K putting.zero /seq-dev/jc18_testcol/
$ ils -L /seq-dev/jc18_testcol/putting.zero
  ap21_testing      0 root;replicate;red;red1;irods-seq-i01-fg            0 2016-07-13.11:24 & putting.zero
    d41d8cd98f00b204e9800998ecf8427e    generic    /irods-seq-i01-fg/jc18_testcol/putting.zero
  ap21_testing      1 root;replicate;green;green6;irods-seq-sr01-ddn-ra08-45-46-47            0 2016-07-13.11:24 & putting.zero
        generic    /irods-seq-sr01-ddn-ra08-45-46-47/jc18_testcol/putting.zero
```

What have I missed?

**Comment by kript** [comment]
its not a permissions issue, AFAIK;

```
$ ils -A /seq-dev/jc18_testcol/ | head
/seq-dev/jc18_testcol:
        ACL - jc18#Sanger1-dev:own   ap21_testing#seq-dev:own   irods-g1-dev#seq-dev:own   
        Inheritance - Disabled
```

**Comment by kript** [comment]
roughly the same occurs when writing to the other path

```
$ time iput -K jc18_deleteme /seq-dev/illumina/library_merge/

real	0m1.975s
user	0m0.036s
sys	0m0.008s
$ ils -L /seq-dev/illumina/library_merge/jc18_deleteme
  jc18              0 demoResc            0 2016-07-13.11:38 & jc18_deleteme
        generic    /usr/local/iRODS/Vault/illumina/library_merge/jc18_deleteme
```

**Comment by trel** [comment]
Can you try setting the second parameter to "forced" for both of these?

```
                msiSetDefaultResc("noReplRoot","null");

                msiSetDefaultResc("root","null");
```

**Comment by kript** [comment]
No change, I'm afraid;

````
$ ienv | grep default
NOTICE: irods_default_resource is not defined
$ time iput -K jc18_deleteme /seq-dev/jc18_testcol/jc18_deleteme4

real	0m1.727s
user	0m0.020s
sys	0m0.004s
$ ils -L !$
ils -L /seq-dev/jc18_testcol/jc18_deleteme4
  jc18              0 demoResc            0 2016-07-14.09:42 & jc18_deleteme4
        generic    /usr/local/iRODS/Vault/jc18_testcol/jc18_deleteme4
```
logs on iCAT
```
Jul 14 09:42:22 pid:10392 NOTICE: Agent process 10566 started for puser=irods-dev and cuser=jc18 from IP_OF_FEDERATION_MASTER
Jul 14 09:42:23 pid:10566 NOTICE: writeLine: inString = writing to root
Jul 14 09:42:23 pid:10392 NOTICE: Agent process 10567 started for puser=irods-g1-dev and cuser=jc18 from IP_OF_FEDERATION_MASTER
Jul 14 09:42:24 pid:10566 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 14 09:42:24 pid:10566 NOTICE: Agent exiting with status = 0
Jul 14 09:42:24 pid:10567 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 14 09:42:24 pid:10567 NOTICE: Agent exiting with status = 0
Jul 14 09:42:24 pid:10392 NOTICE: Agent process 10566 exited with status 0
Jul 14 09:42:24 pid:10392 NOTICE: Agent process 10567 exited with status 0
Jul 14 09:42:28 pid:10392 NOTICE: Agent process 10568 started for puser=irods-dev and cuser=jc18 from IP_OF_FEDERATION_MASTER
Jul 14 09:42:29 pid:10568 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 14 09:42:29 pid:10568 NOTICE: Agent exiting with status = 0
Jul 14 09:42:29 pid:10392 NOTICE: Agent process 10568 exited with status 0
```

no logs on one of the iRES but this on the other;
```
Jul 14 09:42:23 pid:15399 NOTICE: Agent process 11144 started for puser=irods-dev and cuser=jc18 from IP_OF_FEDERATION_MASTER
Jul 14 09:42:24 pid:11144 NOTICE: readAndProcClientMsg: received disconnect msg from client
Jul 14 09:42:24 pid:11144 NOTICE: Agent exiting with status = 0
Jul 14 09:42:24 pid:15399 NOTICE: Agent process 11144 exited with status 0
```

Should the federation master be talking directly to the iRES of a federated zone? 

Aha! one of the iRES was missing the seq-dev.re addition to its server_config.json, now I get the replication working!

So thats two misconfigurations on my part; 
1. not setting the rule to forced
2. not having every iRES have the same ruleset

The first you can do little about if I don't read the docs ;-) However, it might be worth highlighting the behavior of the irods_default_resource, perhaps in the manual on the irods_enviornment.json section?

The second I feel is not clear that you *must* do this, as mostly mention is made of having rules on the iCAT. Obvious in hindsight, but perhaps worth a mention in the docs in the rule section?

Not to hijack this issue, but why does the iRES have two entries from the log, and not one on each iRES?
```
Jul 14 10:15:50 pid:24602 NOTICE: writeLine: inString = writing to root
Jul 14 10:15:50 pid:24602 NOTICE: writeLine: inString = writing to root
```

**Comment by trel** [comment]
This will be fixed by #2713 - this does not appear to be a federation issue, rather, just a lack of policy enforcement.
